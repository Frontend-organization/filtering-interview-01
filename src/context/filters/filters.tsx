import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState
} from 'react'

type FiltersContext = {
  filters: string[]
  setFilters: (newFilters: string[]) => void
}

const filtersContext = createContext<FiltersContext>({} as FiltersContext)

const FiltersProvider: React.FC = ({ children }) => {
  const [filters, _setFilters] = useState<string[]>([])

  const setFilters = useCallback(
    (newFilters: string[] | ((nextFilters: string[]) => string[])) => {
      if (typeof newFilters === 'function') {
        _setFilters(newFilters(filters))
        window.history.replaceState(
          null,
          '',
          `?filters=${newFilters(filters).join('+')}`
        )
      } else {
        _setFilters(newFilters)
        window.history.replaceState(
          null,
          '',
          `?filters=${newFilters.join('+')}`
        )
      }
    },
    [filters]
  )

  useEffect(() => {
    const searchFilters = new URLSearchParams(window.location.search)
    const nextFilters = searchFilters.get('filters')?.split(' ')
    _setFilters(
      nextFilters.filter((f) => f !== '').length > 0 ? nextFilters : []
    )
  }, [])

  return (
    <filtersContext.Provider value={{ filters, setFilters }}>
      {children}
    </filtersContext.Provider>
  )
}

export default FiltersProvider

export const useFilters = (): FiltersContext => useContext(filtersContext)
