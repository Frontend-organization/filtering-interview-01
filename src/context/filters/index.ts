export { default } from './filters'
export * from './filters'
