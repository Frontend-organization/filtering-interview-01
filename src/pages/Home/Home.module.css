body {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen,
    Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  font-size: 62.5%;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
}

.container {
  position: relative;
  height: 100%;
  width: 100%;
}

.navbar-wrapper {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
}

.filter-btn {
  position: fixed;
  left: 50%;
  transform: translateX(-50%);
  bottom: 20px;
}

.fallback {
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 50%;
}

.bottom-selector {
  position: absolute;
  bottom: 0;
  left: 50%;
  transform: translateX(-50%) translateY(0);
  width: 100%;
  max-width: 500px;
  transition: opacity ease-out 0.3s;
}

.btn-close-selector-wrapper {
  position: absolute;
  top: -20px;
  right: 12px;
}

.btn-close-selector {
  background-color: var(--color-white);
  padding: 12px;
}

.bottom-selector.modal-visible {
  animation: appear 0.23s ease-out forwards;
}

@keyframes appear {
  from {
    opacity: 0.5;
    transform: translateX(-50%) translateY(20px);
  }
  to {
    opacity: 1;
    transform: translateX(-50%) translateY(0);
  }
}
