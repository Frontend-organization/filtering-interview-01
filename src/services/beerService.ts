const apiURL = process.env.REACT_APP_DATA_ENDPOINT

export const getBeers = async () => {
  const res = await fetch(apiURL)
  const responseJson = await res.json()
  return responseJson.data
}
