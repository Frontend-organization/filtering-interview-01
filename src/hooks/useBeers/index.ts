import { useEffect, useState } from 'react'
import { getBeers } from '../../services/beerService'

const useBeers = () => {
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState(null)
  const [beers, setBeers] = useState([])

  useEffect(() => {
    setIsLoading(true)
    getBeers()
      .then((data) => {
        setBeers(data)
        console.log(data)
        setIsLoading(false)
      })
      .catch((err) => {
        setError(err)
        console.log(err)
      })
  }, [])

  return { isLoading, beers, error }
}

export default useBeers
