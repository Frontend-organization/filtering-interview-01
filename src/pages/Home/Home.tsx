import CloseIcon from '@assets/icons/close.svg'
import SearchIcon from '@assets/icons/search.svg'
import ShopCartIcon from '@assets/icons/shopping.svg'

import Button from '@components/common/Button'
import Flex from '@components/common/Flex'
import Modal, { ModalRef } from '@components/common/Modal'
import Spacer from '@components/common/Spacer'
import Typography from '@components/common/Typography'
import BeersContainer from '@components/containers/BeersContainer'
import * as FiltersSelector from '@components/containers/FiltersSelector'
import Footer from '@components/containers/Footer'
import Navbar from '@components/containers/Navbar'
import { useFilters } from '@context/filters'
import useBeers from '@hooks/useBeers'
import { Beer } from '@services/beerService'
import { Fragment, useCallback, useMemo, useRef } from 'react'

import s from './Home.module.css'

const { Heading } = Typography
const { FilterSelectorWidget, default: Selector } = FiltersSelector

const Home: React.FC = () => {
  const { filters } = useFilters()
  const { isLoading, beers, error } = useBeers()
  const modalRef = useRef<ModalRef>(null)

  const filteredBeers: Beer[] = useMemo(() => {
    if (filters.length === 0) return beers
    return beers.filter(({ filterId }) => filters.includes(filterId))
  }, [filters, beers])

  const renderContent = useCallback(() => {
    if (isLoading)
      return (
        <div className={s.fallback}>
          <Heading>Cargando...</Heading>
        </div>
      )
    if (error)
      return (
        <div className={s.fallback}>
          <Heading>Ocurri√≥ un error inesperado</Heading>
        </div>
      )
    return (
      <main>
        <BeersContainer beers={filteredBeers} />
      </main>
    )
  }, [error, isLoading, filteredBeers])

  return (
    <section className={s.container}>
      <header className={s['navbar-wrapper']}>
        <Navbar
          logo
          menuButton
          navbarRight={
            <Fragment>
              <SearchIcon width={24} fill="#fff" />
              <Spacer x={16} />
              <ShopCartIcon width={24} fill="#fff" />
            </Fragment>
          }
        />
      </header>
      <Spacer y={80} />
      <Flex justifyContent="center">
        <Heading size="xl" weight="bold">
          Cervezas
        </Heading>
      </Flex>
      {renderContent()}
      <Spacer y={22} />
      <Footer />
      <div className={s['filter-btn']}>
        <FilterSelectorWidget onClick={() => modalRef.current?.show()}>
          Filtrar
        </FilterSelectorWidget>
      </div>
      <Modal ref={modalRef}>
        {({ visible }) => (
          <div
            className={`${s['bottom-selector']} ${
              visible ? s['modal-visible'] : ''
            }`}
          >
            <div className={s['btn-close-selector-wrapper']}>
              <Button
                className={s['btn-close-selector']}
                onClick={() => modalRef.current?.hide()}
              >
                <CloseIcon width={22} />
              </Button>
            </div>
            {visible ? (
              <Selector
                onFilter={() => modalRef.current?.hide()}
                defaultFilters={filters}
              />
            ) : null}
          </div>
        )}
      </Modal>
    </section>
  )
}

export default Home
