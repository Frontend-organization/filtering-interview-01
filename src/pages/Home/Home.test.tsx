import { useFilters } from '@context/filters'
import useBeers from '@hooks/useBeers'
import { Beer } from '@services/beerService'
import { render } from '@testing-library/react'

import Home from '.'

jest.mock('@context/filters')
jest.mock('@hooks/useBeers')

const MockUseFilters = useFilters as jest.Mock
const MockUseBeers = useBeers as jest.Mock

MockUseFilters.mockReturnValue({ filters: [] })
MockUseBeers.mockReturnValue({ beers: [], isLoading: false })

const beers: Beer[] = Array.from({ length: 30 }).map((_, i) => ({
  description: `description-${i}`,
  filterId: `filter-${i}`,
  img: `img-${i}`,
  name: `name-${i}`,
  price: 200
}))

describe('<Home />', () => {
  it('should render without crashing', () => {
    const rendered = render(<Home />)
    expect(rendered).toMatchSnapshot()
  })

  it('should render results', () => {
    const [firstBeer] = beers
    MockUseBeers.mockReturnValueOnce({ beers, isLoading: false })
    const { getByText } = render(<Home />)
    expect(getByText(firstBeer.description)).toBeInTheDocument()
  })

  it('should render loading state', () => {
    MockUseBeers.mockReturnValueOnce({ beers: [], isLoading: true })
    const { getByText } = render(<Home />)
    expect(getByText('Cargando', { exact: false })).toBeInTheDocument()
  })

  it('should render error state', () => {
    MockUseBeers.mockReturnValueOnce({
      beers: [],
      isLoading: false,
      error: 'some-error'
    })
    const { getByText } = render(<Home />)
    expect(getByText('Ocurri√≥ un error', { exact: false })).toBeInTheDocument()
  })
})
