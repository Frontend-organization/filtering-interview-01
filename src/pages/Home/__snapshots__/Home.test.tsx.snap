// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="container"
      >
        <header
          class="navbar-wrapper"
        >
          <section
            class="navbar"
          >
            <div
              class="navbar__left"
            >
              <i
                data-testid="btn-menu"
                fill="#fff"
                height="32"
                width="32"
              />
              <div
                style="height: 0px; width: 10px;"
              />
              <i
                fill="#fff"
                height="30"
                width="43"
              />
            </div>
            <div
              class="navbar__right"
            >
              <i
                fill="#fff"
                width="24"
              />
              <div
                style="height: 0px; width: 16px;"
              />
              <i
                fill="#fff"
                width="24"
              />
            </div>
          </section>
        </header>
        <div
          style="height: 80px; width: 0px;"
        />
        <div
          style="display: flex; justify-content: center;"
        >
          <h1
            class="heading bold xl "
          >
            Cervezas
          </h1>
        </div>
        <main>
          <div
            class="beersContainer"
          />
        </main>
        <div
          style="height: 22px; width: 0px;"
        />
        <footer
          class="Footer"
        >
          <div
            class="Contact"
          >
            <h1
              class="heading bold md "
            >
              ¿Podemos ayudarle?
            </h1>
            <div
              style="height: 18px; width: 0px;"
            />
            <div
              style="display: flex; justify-content: center;"
            >
              <div
                class="BtnFooter"
              >
                <button
                  class="button 
  rounded full secondary Btn"
                  type="button"
                >
                  <i
                    fill="#000"
                    height="28"
                    width="30"
                  />
                </button>
                <p
                  class="paragraph bold md "
                >
                  CHAT
                </p>
              </div>
              <div
                style="height: 0px; width: 110px;"
              />
              <div
                class="BtnFooter"
              >
                <button
                  class="button 
  rounded  secondary Btn"
                  type="button"
                >
                  <i
                    fill="#000"
                    height="28"
                    width="30"
                  />
                </button>
                <p
                  class="paragraph bold md "
                >
                  LLAMAR
                </p>
              </div>
            </div>
          </div>
          <div
            class="Navigation"
          >
            <a
              class="link bold md Link"
              href="/about-us"
            >
              SOBRE NOSOTROS
            </a>
            <a
              class="link bold md Link"
              href="/security-constraints"
            >
              MEDIDAS DE SEGURIDAD (COVID-19)
            </a>
            <a
              class="link bold md Link"
              href="/how-merch"
            >
              CÓMO MERCAR
            </a>
            <a
              class="link bold md Link"
              href="/our-providers"
            >
              PROVEEDORES
            </a>
            <a
              class="link bold md Link"
              href="/contact-us"
            >
              CONTÁCTENOS
            </a>
            <a
              class="link bold md Link"
              href="/support-zone"
            >
              ZONA DE COBERTURA
            </a>
          </div>
          <div
            style="height: 46px; width: 0px;"
          />
          <div
            class="Suscribe"
          >
            <button
              class="button 
  rounded  primary Btn"
              type="button"
            >
              <i
                width="30"
              />
            </button>
            <div
              style="height: 5px; width: 0px;"
            />
            <p
              class="paragraph bold sm "
            >
              POLÍTICA DE TRATAMIENTO DE DATOS PERSONALES
            </p>
          </div>
          <div
            style="height: 36px; width: 0px;"
          />
          <div
            class="CreditsContainer"
          >
            <p
              class="paragraph bold sm Credits"
            >
              © 2020 Axiacore
            </p>
          </div>
        </footer>
        <div
          class="filter-btn"
        >
          <div
            class="widget"
          >
            <button
              class="button 
    primary "
              type="button"
            >
              FILTRAR
              <div
                class="rightElement"
              >
                <i
                  fill="#fff"
                  width="22"
                />
              </div>
            </button>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="container"
    >
      <header
        class="navbar-wrapper"
      >
        <section
          class="navbar"
        >
          <div
            class="navbar__left"
          >
            <i
              data-testid="btn-menu"
              fill="#fff"
              height="32"
              width="32"
            />
            <div
              style="height: 0px; width: 10px;"
            />
            <i
              fill="#fff"
              height="30"
              width="43"
            />
          </div>
          <div
            class="navbar__right"
          >
            <i
              fill="#fff"
              width="24"
            />
            <div
              style="height: 0px; width: 16px;"
            />
            <i
              fill="#fff"
              width="24"
            />
          </div>
        </section>
      </header>
      <div
        style="height: 80px; width: 0px;"
      />
      <div
        style="display: flex; justify-content: center;"
      >
        <h1
          class="heading bold xl "
        >
          Cervezas
        </h1>
      </div>
      <main>
        <div
          class="beersContainer"
        />
      </main>
      <div
        style="height: 22px; width: 0px;"
      />
      <footer
        class="Footer"
      >
        <div
          class="Contact"
        >
          <h1
            class="heading bold md "
          >
            ¿Podemos ayudarle?
          </h1>
          <div
            style="height: 18px; width: 0px;"
          />
          <div
            style="display: flex; justify-content: center;"
          >
            <div
              class="BtnFooter"
            >
              <button
                class="button 
  rounded full secondary Btn"
                type="button"
              >
                <i
                  fill="#000"
                  height="28"
                  width="30"
                />
              </button>
              <p
                class="paragraph bold md "
              >
                CHAT
              </p>
            </div>
            <div
              style="height: 0px; width: 110px;"
            />
            <div
              class="BtnFooter"
            >
              <button
                class="button 
  rounded  secondary Btn"
                type="button"
              >
                <i
                  fill="#000"
                  height="28"
                  width="30"
                />
              </button>
              <p
                class="paragraph bold md "
              >
                LLAMAR
              </p>
            </div>
          </div>
        </div>
        <div
          class="Navigation"
        >
          <a
            class="link bold md Link"
            href="/about-us"
          >
            SOBRE NOSOTROS
          </a>
          <a
            class="link bold md Link"
            href="/security-constraints"
          >
            MEDIDAS DE SEGURIDAD (COVID-19)
          </a>
          <a
            class="link bold md Link"
            href="/how-merch"
          >
            CÓMO MERCAR
          </a>
          <a
            class="link bold md Link"
            href="/our-providers"
          >
            PROVEEDORES
          </a>
          <a
            class="link bold md Link"
            href="/contact-us"
          >
            CONTÁCTENOS
          </a>
          <a
            class="link bold md Link"
            href="/support-zone"
          >
            ZONA DE COBERTURA
          </a>
        </div>
        <div
          style="height: 46px; width: 0px;"
        />
        <div
          class="Suscribe"
        >
          <button
            class="button 
  rounded  primary Btn"
            type="button"
          >
            <i
              width="30"
            />
          </button>
          <div
            style="height: 5px; width: 0px;"
          />
          <p
            class="paragraph bold sm "
          >
            POLÍTICA DE TRATAMIENTO DE DATOS PERSONALES
          </p>
        </div>
        <div
          style="height: 36px; width: 0px;"
        />
        <div
          class="CreditsContainer"
        >
          <p
            class="paragraph bold sm Credits"
          >
            © 2020 Axiacore
          </p>
        </div>
      </footer>
      <div
        class="filter-btn"
      >
        <div
          class="widget"
        >
          <button
            class="button 
    primary "
            type="button"
          >
            FILTRAR
            <div
              class="rightElement"
            >
              <i
                fill="#fff"
                width="22"
              />
            </div>
          </button>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
